File: project/jimmer-spring-boot-starter/src/main/java/org/babyfish/jimmer/spring/SqlClients.java
package org.babyfish.jimmer.spring;

import org.babyfish.jimmer.sql.JSqlClient;
import org.babyfish.jimmer.sql.kt.KSqlClient;
import org.babyfish.jimmer.sql.kt.KSqlClientKt;
import org.jetbrains.annotations.Nullable;
import org.springframework.context.ApplicationContext;

import javax.sql.DataSource;

/**
 * Designed for single data source.
 *
 * <p>However, it is unnecessary to use this class directly
 * because the sqlClient for single data source is created
 * by the jimmer spring boot starter automatically</p>
 */
public class SqlClients {

    public static JSqlClient java(ApplicationContext ctx) {
        return java(ctx, null);
    }

    public static JSqlClient java(ApplicationContext ctx, @Nullable DataSource dataSource) {
        return new JSpringSqlClient(ctx, dataSource, false);
    }

    public static KSqlClient kotlin(ApplicationContext ctx) {
        return kotlin(ctx, null);
    }

    public static KSqlClient kotlin(ApplicationContext ctx, @Nullable DataSource dataSource) {
        return KSqlClientKt.toKSqlClient(
                new JSpringSqlClient(ctx, dataSource, true)
        );
    }
}

File: project/jimmer-spring-boot-starter/src/main/java/org/babyfish/jimmer/spring/SqlClients.kt
package org.babyfish.jimmer.spring

import org.babyfish.jimmer.sql.JSqlClient
import org.babyfish.jimmer.sql.kt.KSqlClient
import org.babyfish.jimmer.sql.kt.cfg.KSqlClientDsl
import org.babyfish.jimmer.sql.kt.toKSqlClient
import org.springframework.context.ApplicationContext
import java.util.function.Consumer
import javax.sql.DataSource

object SqlClients {

    @JvmStatic
    fun java(
        ctx: ApplicationContext
    ): JSqlClient =
        java(ctx, null, null)

    @JvmStatic
    fun java(
        ctx: ApplicationContext,
        dataSource: DataSource?
    ): JSqlClient =
        java(ctx, dataSource, null)

    @JvmStatic
    fun java(
        ctx: ApplicationContext,
        block: Consumer<JSqlClient.Builder>?
    ): JSqlClient =
        java(ctx, null, block)

    @JvmStatic
    fun java(
        ctx: ApplicationContext,
        dataSource: DataSource?,
        block: Consumer<JSqlClient.Builder>?
    ): JSqlClient =
        JSpringSqlClient(ctx, dataSource, block, false)

    @JvmStatic
    fun kotlin(ctx: ApplicationContext): KSqlClient =
        kotlin(ctx, null, null)

    @JvmStatic
    fun kotlin(
        ctx: ApplicationContext,
        dataSource: DataSource?
    ): KSqlClient =
        kotlin(ctx, dataSource, null)

    @JvmStatic
    fun kotlin(
        ctx: ApplicationContext,
        block: (KSqlClientDsl.() -> Unit)?
    ): KSqlClient =
        kotlin(ctx, null, block)

    @JvmStatic
    fun kotlin(
        ctx: ApplicationContext,
        dataSource: DataSource?,
        block: (KSqlClientDsl.() -> Unit)?
    ): KSqlClient =
        JSpringSqlClient(
            ctx,
            dataSource,
            block?.let {
                Consumer {
                    KSqlClientDsl(it).block()
                }
            },
            true
        ).toKSqlClient()
}